include/matx/core/tensor.h

```diff
@@ -896,6 +906,47 @@ class tensor_t : public detail::tensor_impl_t<T,RANK,Desc> {
     Desc new_desc{this->desc_.Shape(), std::move(strides)};  
     return tensor_t<Type, RANK, Storage, Desc>{storage_, std::move(new_desc), data};
   }
+  
+  /**
+   * Permute the dimensions of a tensor
+   *
+   * Accepts any order of permutation. Number of dimensions must match RANK of
+   * tensor
+   *
+   * @tparam M
+   *   Rank of tensor to permute. Should not be used directly
+   *
+   * @param dims
+   *   Dimensions of tensor
+   *
+   * @returns tensor view of only imaginary-valued components
+   *
+   */
+  __MATX_INLINE__ auto Permute(const std::array<int32_t, RANK> &dims) const
+  {
+    MATX_NVTX_START("", matx::MATX_NVTX_LOG_API)
+    
+    static_assert(RANK >= 2, "Only tensors of rank 2 and higher can be permuted.");
+    std::array<shape_type, RANK> n;
+    std::array<stride_type, RANK> s;
+    [[maybe_unused]] bool done[RANK] = {0};
+
+#pragma unroll
+    for (int i = 0; i < RANK; i++) {
+      int d = dims[i];
+      MATX_ASSERT_STR(d < RANK, matxInvalidDim,
+                      "Index to permute is larger than tensor rank");
+      MATX_ASSERT_STR(done[d] == false, matxInvalidParameter,
+                      "Cannot list the same dimension to permute twice");
+      done[d] = true;
+      n[i] = this->Size(d);
+      s[i] = this->Stride(d);
+    }
+
+    Desc new_desc{std::move(n), std::move(s)};  
+    return tensor_t<T, RANK, Storage, Desc>{storage_, std::move(new_desc), this->ldata_};
+  }
+
```

Comment:
The C version below should be changed to use this one to avoid code duplication

Reply:
I tried that but there apparently is a sign change from uint to int and I didn't want to include a fix for that.  I think we probably need to have 2 APIs one for signed and unsigned unfortunately.

Reply:
__MATX_INLINE__ auto Permute(const std::array<int32_t, RANK> &dims) const
 __MATX_INLINE__ auto Permute(const uint32_t (&dims)[RANK]) const

Second is unsigned.  We can change that to signed but then some of our examples/uses were not compiling.  I'm guessing we are hitting some odd c++ rules about literals being signed vs unsigned.

Reply:
so the second above cannot call detail::to_array as it would produce a uint32_t array.

Reply:
ok changing to int fixed.  I don't think we need a uint version.